### PR TITLE
rgw: fix accessing expired memory in PrefixableSignatureHelper.

### PR DESCRIPTION
**Pull request by rzarzynski**
Fixes: http://tracker.ceph.com/issues/21085
Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

### CODE REVIEW

File: src/rgw/rgw_swift_auth.cc [hunk 1]
```diff
@@ -210,8 +210,9 @@ class TempURLEngine::PrefixableSignatureHelper
       object_name(object_name),
       prefix(prefix) {
     /* Transform: v1/acct/cont/obj - > v1/acct/cont/ */
-    no_obj_uri = \
-      decoded_uri.substr(0, decoded_uri.length() - object_name.length());
+    const size_t no_obj_url_len = \
+      this->decoded_uri.length() - object_name.length();
+    no_obj_uri = this->decoded_uri.substr(0, no_obj_url_len);
```
Comment (cbodley) [inline src/rgw/rgw_swift_auth.cc]: looks right. but renaming the constructor argument so it doesn't shadow the member variable may be preferable

Comment (rzarzynski) [inline src/rgw/rgw_swift_auth.cc]: Reworked.

**Comment by rzarzynski** [comment]
The patch has been tested with Swift's functional tests:
```
http://qa-proxy.ceph.com/teuthology/rzarzynski-2017-08-23_15:23:33-rgw-wip-rzarzynski-testing---basic-mira/1555012/remote/mira041/log/valgrind/cep

$ SWIFT_TEST_CONFIG_FILE=rgw.conf nosetests -v test.functional.test_tempurl:TestTempURLPrefix
...
SKIPPING FUNCTIONAL TESTS SPECIFIC TO AUTH VERSION 3
SKIPPING FUNCTIONAL TESTS SPECIFIC TO SERVICE TOKENS
SKIPPING FUNCTIONAL TESTS DUE TO NO CONFIG FOR RESELLER ADMIN
SKIPPING FUNCTIONAL TESTS SPECIFIC TO AUTH VERSION 3
SKIPPING FUNCTIONAL TESTS SPECIFIC TO SERVICE TOKENS
SKIPPING FUNCTIONAL TESTS DUE TO NO CONFIG FOR RESELLER ADMIN
test_GET (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_GET_DLO_inside_container (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_GET_DLO_outside_container (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_GET_with_key_2 (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_HEAD (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_PUT (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_PUT_manifest_access (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_changing_expires (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_changing_sig (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_different_object (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_empty_prefix (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_missing_query_parm (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_no_prefix_match (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_object_url_with_prefix (test.functional.test_tempurl.TestTempURLPrefix) ... ok

----------------------------------------------------------------------
Ran 14 tests in 1.606s

OK
```

**Comment by yuriw** [comment]
http://pulpito.front.sepia.ceph.com:80/yuriw-2017-08-29_19:33:51-rgw-wip-yuri-testing4_2017_8_30-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-08-30_02:10:47-rgw-wip-yuri-testing4_2017_8_30-distro-basic-smithi/

**Pull request merged**